src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala

```diff
@@ -141,20 +143,66 @@ class AppsResource @Inject() (
     }
   }
 
+  /**
+    * Validate and normalize a single application update submitted via the REST API. Validation exceptions are not
+    * handled here, that's left as an exercise for the caller.
+    *
+    * @param appId used as the id of the generated app update (vs. whatever might be in the JSON body)
+    * @param body is the raw, unparsed JSON
+    * @param partialUpdate true if the JSON should be parsed as a partial application update (all fields optional)
+    *                      or as a wholesale replacement (parsed like an app definition would be)
+    */
+  def canonicalAppUpdateFromJson(appId: PathId, body: Array[Byte], partialUpdate: Boolean): AppUpdate = {
+    if (partialUpdate) {
+      validateOrThrow(Json.parse(body).as[AppUpdate].copy(id = Some(appId)).withCanonizedIds())
+    } else {
+      // this is a complete replacement of the app as we know it, so parse and normalize as if we're dealing
+      // with a brand new app because the rules are different (for example, many fields are non-optional with brand-new apps).
+      // however since this is an update, the user isn't required to specify an ID as part of the definition so we do
+      // some hackery here to pass initial JSON parsing.
+      val jsObj = Json.parse(body).as[JsObject] + ("id" -> Json.toJson(appId.toString))
+      // the version is thrown away in toUpdate so just pass `zero` for now
+      normalizedApp(validateOrThrow(jsObj.as[AppDefinition].copy(id = appId).withCanonizedIds()), Timestamp.zero).toUpdate
```

Comment:
we don't actually need to copy the id in this step since we just did it in the previous one